snuba/settings/__init__.py

```diff
@@ -38,7 +38,7 @@
     str(x): 0 for x in range(0, SENTRY_LOGICAL_PARTITIONS)
 }
 # Storage names to apply dataset partitioning to
-PARTITIONED_STORAGES: Set[str] = set()
+PARTITIONED_STORAGE_SETS: Set[str] = {"migrations", "generic_metrics_sets"}
```

Comment:
Is this config specific to the migration framework or is it used by snuba in other places. 
I think we should try to keep them separated as we want to see the migrations tool as a stand alone entity that we could break out of snuba. 
Right now it only depends on Storage set keys and storage keys. Those are not going to be enums anymore soon so no big deal, though we should try to avoid the logical dependencies as well.

Reply:
I was assuming `StorageSetKey.at_partition` would be used in the query path to switch between storage sets depending on logical partition (though logical : physical mapping isn't yet being used):

https://github.com/getsentry/snuba/pull/3148/files#diff-045703230a5e4a12e1bde538fccabfab59d530d2bf982f8b21c94ab0d8010ddbR79-R87

Reply:
I am not sure migrations can be partitioned at storage set level (allowing a migrations group to contain both partitioned and not partitioned migrations).
The migration tool enforces that, for a migration table, migrations for the same group are executed in sequence (you cannot skip them). 
Though if a migration group contains migrations for multiple storage sets (which I think is possible) you have a problem:
- let's assume group `A` contains migrations from storage sets `B` and `C`.  Now `B` is partitioned on 5 slices while `C` has no slices. 
- You will not have problem in running migrations on the table for slice 0. All Storage sets are present there. 
- The table for slice 1 will become inconsistent as it will be missing the migrations from storage set `C`. The tool will fail there. 

So I think we need to revisit this.
Either allow groups to skip migrations, or disallow different storage sets in the same group.
Or make slicing not tied to the storage set but to something else.

I think this may become controversial.

Reply:
I think this whole system gets incredibly complicated if we allow for different levels of physical slicing (at a settings/conceptual level) within the same deployment. So if we have n physical partitions (n << 256) I think those partitions need to apply across all CH storages *even if* some storages will internally map multiple physical partitions to one physical CH cluster.

Example:
There are 4 physical partitions in a snuba deploy.
`generic_metrics_distributions`, which contains the most metrics data, maps 
```
{
    physical_part_1: cluster_1,
    physical_part_2: cluster_2,
    physical_part_3: cluster_3,
    physical_part_4: cluster_4
}
```

whereas `generic_metrics_sets` which doesn't require nearly as much space will map
```
{
    physical_part_1: cluster_1,
    physical_part_2: cluster_1,
    physical_part_3: cluster_1,
    physical_part_4: cluster_1
}
```

It's sort of like a second level of logical mapping, but it means we can reuse the migrations table for all the physical partitions in the second group but run them independently for the first.

(I'm not taking into consideration Kafka here)

Reply:
I am not sure we are addressing the same problem. 
My concern is that having a migrations table per slice breaks the migration tools data model. 
- The migration tool data model require sequentiality within a migration group. You cannot skip migrations.
- There is nothing forcing a group to contain migrations that only reference one storage key/set 
- Thus, the same group can contain at the same time migrations for a sliced storage and migrations for a non sliced storage.
- We would not like to change any of the constraint above

This implies, that running slice 2 on a separate migration table may require you to skip migrations in a group as the group may contain migrations that should not be applied to slice 2 as they are for a non sliced storage.
Skipping migrations in a group breaks the sequentiality point above.